### PR TITLE
Record Dot Hover Types

### PR DESCRIPTION
**Pull request by coltenwebb**
Working towards #2732 (reopened since I closed and force-pushed my last PR, which github dislikes)

This PR patches the way HieAST is generated by GHC. For performance reasons, not all nodes in the AST get expanded and checked for types. This caused the nodes corresponding to record-dot-syntax to not be included in the HieAST. This patch adds them back.

## Notes

This is what type info on hover looks like:

Code_ps0jgEHy0v

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3016"></a>

### CODE REVIEW

**Comment by kokobd** [comment]
Do we have to introduce the large file with 2000+ lines? It looks similar to `GHC/Iface/Ext/Ast.hs`, could you explain it a bit?

**Comment by coltenwebb** [comment]
You're right, the `HieAst.hs` in src-ghc92 is the same one in 9.2.x GHC except I inserted the line 789, to ensure the record dot syntax gets desugared. I agree that adding so much to the codebase is less than ideal. Maybe this explanation belongs in a banner comment at the top of the file? I'll also try to get this patch directly into GHC, but it may be better to keep this here for backwards compatibility with 9.2.{1,2}

**Comment by michaelpj** [comment]
I think that's the nature of the `hie-compat` packages, they mostly consist of wholesale copies of files from GHC with small changes. I'm not sure if we can easily improve that, but comments are _definitely_ a good idea.

**Comment by wz1000** [comment]
We would want a patch that is accepted for GHC HEAD before we can merge such changes to hie-compat. We have followed this policy for all the previous patches we have in `hie-compat`.

**Comment by coltenwebb** [comment]
Should be good to go. I've added the comments to HieAst.hs, and the patch is merged into GHC master.

**Comment by wz1000** [comment]
Could you add the patch to the readme of hie-compat?

**Review by wz1000 — APPROVED** [review]
Thanks, looks all good now.

**Comment by michaelpj** [comment]
Just needs a test, I think?

**Comment by coltenwebb** [comment]
I added three tests, there are additional tests for the record dot syntax types on the GHC side too.

**Comment by coltenwebb** [comment]
✨

**Comment by michaelpj** [comment]
Sweet!

File: ghcide/test/exe/Main.hs [hunk 1]
```diff
@@ -4352,16 +4352,25 @@ findDefinitionAndHoverTests = let
...
-  typeDefinitionTests = [ tst (getTypeDefinitions, checkDefs) aaaL14 (pure tcData) "Saturated data con"
-                        , tst (getTypeDefinitions, checkDefs) aL20 (pure [ExpectNoDefinitions]) "Polymorphic variable"]
+    , testGroup "type-definition" typeDefinitionTests
+    , testGroup "hover-record-dot-syntax" recordDotSyntaxTests ]
+
+  typeDefinitionTests = [ tst (getTypeDefinitions, checkDefs) aaaL14 sourceFilePath (pure tcData) "Saturated data con"
+                        , tst (getTypeDefinitions, checkDefs) aL20 sourceFilePath (pure [ExpectNoDefinitions]) "Polymorphic variable"]
+  
+  recordDotSyntaxTests 
+    | ghcVersion == GHC92 =
```
Comment (michaelpj) [inline ghcide/test/exe/Main.hs:4362]: Ah, too late: should this check be more generous? We expect these tests to work on later versions also, right?

Comment (coltenwebb) [reply ghcide/test/exe/Main.hs:4362]: Oh that's correct. I'll make a PR for that

**Pull request merged**